File: backend/app/public/routes/bookings.py
from uuid import UUID

from fastapi import APIRouter, Depends

from app.api.response import ApiResponse
from app.auth.jwt.context import TokenContext
from app.dependencies.auth import require_permission
from app.dependencies.rls import get_rls_db
from app.dependencies.service import get_service
from app.public.services.bookings_create import CreateBookingService
from app.public.services.bookings_read import ReadBookingService
from app.public.schemas.booking import (
    CreateBookingRequest,
    CreateBookingResponse,
    BookingResponse,
    BookingTableResponse,
    ListBookingsFilterRequest,
    BookingConflictRow,
)


router = APIRouter(
    prefix="/booking/v1",
    tags=["booking"],
)


@router.post("/", status_code=201)
def create_booking(
    body: CreateBookingRequest,
    ctx: TokenContext = Depends(
        require_permission(
            resource="bookings",
            action="c"
        )
    ),
    service: CreateBookingService = Depends(
        get_service(
            CreateBookingService,
            db_dependency=get_rls_db,
        )
    ),
) -> ApiResponse[CreateBookingResponse]:
    result = service.create(
        body=body,
        user_id=ctx.user_id,
        location_id=ctx.location_id
    )

    return ApiResponse(
        ok=True,
        msg="Created Pending Bookings",
        data=result
    )


@router.post("/conflicts")
def find_conflicts(
    body: CreateBookingRequest,
    ctx: TokenContext = Depends(
        require_permission(
            resource="bookings",
            action="c",
        )
    ),
    service: CreateBookingService = Depends(
        get_service(
            CreateBookingService,
            db_dependency=get_rls_db,
        )
    ),
) -> ApiResponse[list[BookingConflictRow]]:
    result = service.find_conflicts(
        body=body,
    )

    return ApiResponse(
        ok=True,
        msg="Checked Booking Conflicts",
        data=result,
    )


@router.get("/{booking_id}")
def get_booking(
    booking_id: UUID,
    ctx: TokenContext = Depends(
        require_permission(
            resource="bookings",
            action="r"
        )
    ),
    service: ReadBookingService = Depends(
        get_service(
            ReadBookingService,
            db_dependency=get_rls_db,
        )
    ),
) -> ApiResponse[BookingResponse]:
    result = service.get_by_id(
        id=booking_id
    )

    return ApiResponse(
        ok=True,
        msg="Fetched Booking",
        data=result
    )


@router.get("/")
def list_bookings(
    filters: ListBookingsFilterRequest = Depends(),
    ctx: TokenContext = Depends(
        require_permission(
            resource="bookings",
            action="r"
        )
    ),
    service: ReadBookingService = Depends(
        get_service(
            ReadBookingService,
            db_dependency=get_rls_db,
        )
    ),
) -> ApiResponse[list[BookingTableResponse]]:
    result = service.list_with_filters(
        filters=filters
    )

    return ApiResponse(
        ok=True,
        msg="Fetched Bookings",
        data=result
    )

File: frontend/src/lib/api/public/adapters/bookings.ts
import { z } from 'zod';
import { PUBLIC_API_URL } from '$env/static/public';

import { apiWrapper } from '$lib/api/wrapper';
import { buildQueryParams } from '$lib/api/param';

import type { Pagination } from '$lib/types/pagination';

import {
	createBookingResponseSchema,
	bookingConflictRowSchema,
	bookingResponseSchema,
	bookingTableResponseSchema,
	type CreateBookingRequest,
	type CreateBookingResponse,
	type BookingConflictRow,
	type BookingResponse,
	type BookingTableResponse,
	type ListBookingsFilterRequest
} from '../types/bookings';


export function createBookingApi(
	fetcher?: typeof fetch
) {
	const baseUrl =
		`${PUBLIC_API_URL}/booking/v1`;

	return {
		create(body: CreateBookingRequest) {
			return apiWrapper<CreateBookingResponse>(
				`${baseUrl}/`,
				{
					method: 'POST',
					body,
					fetcher,
					schema: createBookingResponseSchema
				}
			);
		},

		findConflicts(body: CreateBookingRequest) {
			return apiWrapper<BookingConflictRow[]>(
				`${baseUrl}/conflicts`,
				{
					method: 'POST',
					body,
					fetcher,
					schema: z.array(bookingConflictRowSchema)
				}
			);
		},

		getById(bookingId: string) {
			return apiWrapper<BookingResponse>(
				`${baseUrl}/${bookingId}`,
				{
					method: 'GET',
					fetcher,
					schema: bookingResponseSchema
				}
			);
		},

		list(
			filters: ListBookingsFilterRequest = {},
			pagination: Pagination = {}
		) {
			const query = buildQueryParams(
				filters,
				pagination
			);

			return apiWrapper<BookingTableResponse[]>(
				`${baseUrl}/${query ? `?${query}` : ''}`,
				{
					method: 'GET',
					fetcher,
					schema: z.array(bookingTableResponseSchema)
				}
			);
		}
	};
}